booster/library/Booster/Pattern/Implies.hs

```diff
@@ -108,18 +111,20 @@ runImplies def mLlvmLibrary mSMTOptions antecedent consequent =
                                 -- apply the given substitution before doing anything else
                                 substPatL =
                                     Pattern
-                                        { term = substituteInTerm substitutionL patL.term
-                                        , constraints = Set.map (substituteInPredicate substitutionL) patL.constraints
+                                        { term = Substitution.substituteInTerm substitutionL patL.term
+                                        , constraints = Set.map (Substitution.substituteInPredicate substitutionL) patL.constraints
                                         , ceilConditions = patL.ceilConditions
+                                        , substitution = substitutionL
                                         }
                                 substPatR =
                                     Pattern
-                                        { term = substituteInTerm substitutionR patR.term
-                                        , constraints = Set.map (substituteInPredicate substitutionR) patR.constraints
+                                        { term = Substitution.substituteInTerm substitutionR patR.term
+                                        , constraints = Set.map (Substitution.substituteInPredicate substitutionR) patR.constraints
                                         , ceilConditions = patR.ceilConditions
+                                        , substitution = substitutionR
                                         }
```

Comment:
We very often construct a pattern directly, bypassing the `Pattern_` smart-constructor. I'd love to have a smart constructor that is used consistently, but I think it may be out-of-scope for this PR. I do agree that it's best to apply the substitution in the internalisation function that return a `Pattern`.